### PR TITLE
[tune] Update pytorch-lightning integration API

### LINKED ISSUE #38837: [train][GA] Update `tune.integration.pytorch_lightning`
Same thing as xgboost/lgbm updates. 

- [ ] Hard deprecate `TuneReportCallback`
- [ ] Hard deprecate `TuneCheckpointCallback`
- [ ] Update `TuneReportCheckpointCallback` to use `train.report`

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Updates PTL callbacks to use new train.report API.

## Related issue number

Closes https://github.com/ray-project/ray/issues/38837

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/integration/pytorch_lightning.py [hunk 1]
```diff
@@ -146,102 +154,58 @@ def _get_report_dict(self, trainer: Trainer, pl_module: LightningModule):
...
+            return
 
+        with tempfile.TemporaryDirectory() as checkpoint_dir:
+            trainer.save_checkpoint(os.path.join(checkpoint_dir, self._filename))
 
-    """
+            if _use_storage_context():
+                checkpoint = Checkpoint.from_directory(checkpoint_dir)
+            else:
+                checkpoint = LegacyCheckpoint.from_directory(checkpoint_dir)
```
Comment (justinvyu) [inline python/ray/tune/integration/pytorch_lightning.py:171]: we can just use `train.Checkpoint` now -- the alias is now gated by the feature flag.

File: python/ray/tune/integration/pytorch_lightning.py [hunk 2]
```diff
@@ -146,102 +154,58 @@ def _get_report_dict(self, trainer: Trainer, pl_module: LightningModule):
...
+        with self._get_checkpoint(trainer) as checkpoint:
+            train.report(report_dict, checkpoint=checkpoint)
 
 
-    """
+class _TuneCheckpointCallback(TuneCallback):
+    def __init__(self, *args, **kwargs):
+        raise DeprecationWarning(
+            "`ray.tune.integration.pytorch_lightning._TuneCheckpointCallback` "
+            "is deprecated."
```
Comment (justinvyu) [inline python/ray/tune/integration/pytorch_lightning.py:191]: nit: Use TuneReportCheckpointCallback instead

File: python/ray/tune/integration/pytorch_lightning.py [hunk 3]
```diff
@@ -146,102 +154,58 @@ def _get_report_dict(self, trainer: Trainer, pl_module: LightningModule):
...
-
-    def _handle(self, trainer: Trainer, pl_module: LightningModule):
-        self._checkpoint._handle(trainer, pl_module)
-        self._report._handle(trainer, pl_module)
+        if log_once("tune_ptl_report_deprecated"):
+            warnings.warn(
+                "`ray.tune.integration.pytorch_lightning.TuneReportCallback` "
+                "is deprecated. Use "
+                "`ray.tune.integration.pytorch_lightning.TuneCheckpointReportCallback`"
+                " instead."
```
Comment (justinvyu) [inline python/ray/tune/integration/pytorch_lightning.py:207]:
```suggestion
                "`ray.tune.integration.pytorch_lightning.TuneReportCheckpointCallback`"
                " instead."
```

**Review by justinvyu — APPROVED** [review]
Thanks! some nits

**Comment by zhe-thoughts** [comment]
@matthewdeng can I have your approval since it's late stage. Thanks

**Review by matthewdeng — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
Please merge into master first and create cherry pick PR. Thanks

**Pull request merged**